sdk/metric/aggregator/histogram/histogram.go

```diff
@@ -92,58 +116,82 @@ func New(desc *export.Descriptor, boundaries []core.Number) *Aggregator {
 
 // Sum returns the sum of all values in the checkpoint.
 func (c *Aggregator) Sum() (core.Number, error) {
-	return c.checkpoint.sum, nil
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.checkpoint().sum, nil
 }
 
 // Count returns the number of values in the checkpoint.
 func (c *Aggregator) Count() (int64, error) {
-	return int64(c.checkpoint.count.AsUint64()), nil
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return int64(c.checkpoint().count), nil
 }
 
 // Histogram returns the count of events in pre-determined buckets.
 func (c *Aggregator) Histogram() (aggregator.Buckets, error) {
-	return c.checkpoint.buckets, nil
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.checkpoint().buckets, nil
 }
 
 // Checkpoint saves the current state and resets the current state to
 // the empty set.  Since no locks are taken, there is a chance that
 // the independent Sum, Count and Bucket Count are not consistent with each
 // other.
 func (c *Aggregator) Checkpoint(ctx context.Context, desc *export.Descriptor) {
-	// N.B. There is no atomic operation that can update all three
-	// values at once without a memory allocation.
-	//
-	// This aggregator is intended to trade this correctness for
-	// speed.
-	//
-	// Therefore, atomically swap fields independently, knowing
-	// that individually the three parts of this aggregation could
-	// be spread across multiple collections in rare cases.
-
-	c.checkpoint.count.SetUint64(c.current.count.SwapUint64Atomic(0))
-	c.checkpoint.sum = c.current.sum.SwapNumberAtomic(core.Number(0))
-
-	for i := 0; i < len(c.checkpoint.buckets.Counts); i++ {
-		c.checkpoint.buckets.Counts[i].SetUint64(c.current.buckets.Counts[i].SwapUint64Atomic(0))
-	}
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	// clear the checkpoint state before making it the new current state.
+	c.resetCheckpoint()
+
+	// swap the current and checkpoint state by changing the least significant bit
+	// of the currentIdx
+	atomic.AddUint32(&c.currentIdx, 1)
+
+	// TODO(paivagustavo): there is a chance where the SDK calls Checkpoint and an update
```

Comment:
@jmacd could you please take a look at this PR before I make it ready for other reviewers? Specially this part, if this is not okay I may need to do work some more on this.